core/src/main/scala/cats/data/NonEmptyList.scala

```diff
@@ -0,0 +1,159 @@
+   */
+  def forall(p: A => Boolean): Boolean =
+    p(head) && tail.exists(p)
+
+  /**
+   * Left-associative fold on the structure using f.
+   */
+  def foldLeft[B](b: B)(f: (B, A) => B): B =
+    (head :: tail).foldLeft(b)(f)
+
+  /**
+   * Right-associative fold on the structure using f.
+   */
+  def foldRight[B](lb: Eval[B])(f: (A, Eval[B]) => Eval[B]): Eval[B] =
+    (head :: tail).foldRight(lb)(f)
+
+  /**
+   *  Applies f to all the elements of the structure
+   */
+  def map[B](f: A => B): NonEmptyList[B] =
+    NonEmptyList(f(head), tail.map(f))
+}
+
+private[data] sealed trait NonEmptyListInstances extends NonEmptyListLowPriority2 {
+
+  implicit def catsDataEqForNonEmptyList[A](implicit A: Eq[A]): Eq[NonEmptyList[A]] =
+    new Eq[NonEmptyList[A]]{
+      def eqv(x: NonEmptyList[A], y: NonEmptyList[A]): Boolean = x === y
+    }
+
+  implicit def catsDataShowForNonEmptyList[A](implicit A: Show[A]): Show[NonEmptyList[A]] =
+    Show.show[NonEmptyList[A]](_.show)
+
+  implicit def catsDataSemigroupKForNonEmptyList[A]: SemigroupK[NonEmptyList[?]] =
+    new SemigroupK[NonEmptyList[?]] {
+      def combineK[A](a: NonEmptyList[A], b: NonEmptyList[A]): NonEmptyList[A] =
+        a combine b
+    }
+
+  implicit def catsDataSemigroupForNonEmptyList[A]: Semigroup[NonEmptyList[A]] =
+    catsDataSemigroupKForNonEmptyList[F].algebra
+
+  implicit def catsDataReducibleForNonEmptyList[A]: Reducible[NonEmptyList[?]] =
+    new NonEmptyReducible[NonEmptyList[?]] {
+      override def split[A](fa: NonEmptyList[A]): (A, F[A]) = (fa.head, fa.tail)
+    }
+
+  implicit def catsDataMonadForNonEmptyList[A]: Monad[NonEmptyList[?]] =
+    new Monad[NonEmptyList[?]] {
+      override def map[A, B](fa: NonEmptyList[F, A])(f: A => B): NonEmptyList[F, B] =
+        fa map f
+
+      def pure[A](x: A): NonEmptyList[F, A] =
+        NonEmptyList(x, monad.empty)
+
+      def flatMap[A, B](fa: NonEmptyList[F, A])(f: A => NonEmptyList[F, B]): NonEmptyList[F, B] = {
+        val end = monad.flatMap(fa.tail) { a =>
+          val fa = f(a)
+          monad.combineK(monad.pure(fa.head), fa.tail)
+        }
+        val fst = f(fa.head)
+        NonEmptyList(fst.head, monad.combineK(fst.tail, end))
+      }
+    }
+}
+
+trait NonEmptyListLowPriority0 {
+  implicit val nelComonad: Comonad[NonEmptyList[List, ?]] =
+    new Comonad[NonEmptyList[List, ?]] {
+      def coflatMap[A, B](fa: NonEmptyList[List, A])(f: NonEmptyList[List, A] => B): NonEmptyList[List, B] = {
+        @tailrec def consume(as: List[A], buf: ListBuffer[B]): List[B] =
+          as match {
+            case Nil => buf.toList
+            case a :: as => consume(as, buf += f(NonEmptyList(a, as)))
+          }
+        NonEmptyList(f(fa), consume(fa.tail, ListBuffer.empty))
+      }
+
+      def extract[A](fa: NonEmptyList[List, A]): A =
+        fa.head
+
+      def map[A, B](fa: NonEmptyList[List, A])(f: A => B): NonEmptyList[List, B] =
+        fa map f
+    }
+}
```

Comment:
For `OneAnd`, where `F` is abstract, you have to break down a bunch of separate instances. Since we are now specializing to `List`, I think we can probably get by without creating intermediate traits like `NonEmptyListFunctor`. I think we can probably have a bunch of instances smashed into one like we do [here](https://github.com/typelevel/cats/blob/5a38209b2a13ded2652528a8767b1683e74d339d/core/src/main/scala/cats/std/option.scala#L9) for `List` (only in this case we will also have `Comonad`, `Reducible`, etc.